### ISSUE TITLE
`local -i` or `declare -i` with quoting

### ISSUE THREAD

**Issue by pihentagy** [koalaman/shellcheck#1156]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2086
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

<details>
<summary>For new checks and feature suggestions (not relevant)</summary>
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related
</details>

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

declare -i z
z=$(some calc)
any_command $z
```

#### Here's what shellcheck currently says:
```
$ shellcheck myscript
 
Line 5:
any_command $z
     ^-- SC2086: Double quote to prevent globbing and word splitting.

$ 
```

#### Here's what I wanted or expected to see:

\<no warnings\>

**Comment by Dmole** [comment]
If shellcheck becomes type aware

A=$(find . | grep -c)
cat $A

etc should also be permitted.

**Comment by koalaman** [comment]
ShellCheck already doesn't warn for e.g.

    var=42
    echo $var

for the same reason. Unfortunately, this analysis is extremely naive, and mostly intended for global constants. With proper dataflow analysis, this kind of thing should be easy to add.

That's not the case for `grep -c` though, since output will be empty for invalid regex or fork failures.

**Comment by mpu** [comment]
I'd be interested in implementing a fix for the issue. But before I start any work, @koalaman do you think it'd be acceptable to skip quoting warnings for variables which we know were declared with the -i flag?

From some trivial experiments it looks like it could be safe:
```sh
#!/bin/bash
declare -i x='abc'
echo $x # shows 0
x='"foo"' # bash syntax error
```

**Comment by koalaman** [comment]
It should be. Realistically it'll probably be ok to suppress the warning for all `declare -i` names in general, and there's already precedence for doing things like that if you want to give it a go.

It won't work perfectly for e.g.:

```
x="*"
echo $x  # This isn't an integer
declare -i x="1"
echo $x  # This is
```

but fixing that is a more significant undertaking.

**Comment by mkoskar** [comment]
Consider following 3 assignments which are equivalent (we could argue about which is better style-wise) but aren't treated the same.

```shell
#!/usr/bin/bash

declare -i x=1

#x=$((OPTARG))  # ok
#x=OPTARG       # ok

x=$OPTARG       # triggers SC2086 on echo below

echo $x

```

**Comment by koalaman** [comment]
This significant undertaking was undertook in #2541, and `declare -i` should now correctly suppress quotes in the appropriate scope.
